app/workers/travis_branch_monitor.rb

```diff
@@ -0,0 +1,104 @@
+require 'travis'
+
+class TravisBranchMonitor
+  include Sidekiq::Worker
+  sidekiq_options :queue => :miq_bot_glacial, :retry => false
+
+  include Sidetiq::Schedulable
+  recurrence { hourly.minute_of_hour(0, 15, 30, 45) }
+
+  include SidekiqWorkerMixin
+
+  class << self
+    private
+
+    # For this class, repos will sometimes the repo needs to be mapped to a
+    # specific gitter room, so a hash is required.
+    #
+    # This override allows for doing this in the config
+    #
+    #   travis_branch_monitor:
+    #     included_repos:
+    #       ManageIQ/manageiq-ui-classic: ManageIQ/ui
+    #       ManageIQ/manageiq-gems-pending: ManageIQ/core
+    #       ManageIQ/manageiq:
+    #       ManageIQ/miq_bot:
+    #
+    # Which you are allowed to leave the value empty, and the key will be used
+    # where appropriate (not used in this class).
+    #
+    # The result from the above for this method will then be:
+    #
+    #   [
+    #     "ManageIQ/manageiq-ui-classic",
+    #     "ManageIQ/manageiq-gems-pending",
+    #     "ManageIQ/manageiq",
+    #     "ManageIQ/miq_bot"
+    #   ]
+    #
+    def included_and_excluded_repos
+      super # just used for error handling...
+
+      [
+        settings.included_repos.try(:to_h).try(:stringify_keys).try(:keys),
+        settings.excluded_repos.try(:to_h).try(:stringify_keys).try(:keys)
+      ]
+    end
+  end
+
+  def perform
+    if !first_unique_worker?
+      logger.info("#{self.class} is already running, skipping")
+    else
+      process_repos
+    end
+  end
+
+  def process_repos
+    enabled_repos.each do |repo|
+      process_repo(repo)
+    end
+  end
+
+  def process_repo(repo)
+    # repo.regular_branch_names.each do |branch|
+    # end
+
+    # TODO:  Handle this for all release branches
+    process_branch(repo, 'master')
+  end
+
+  def process_branch(repo, branch)
+    branch_record = repo.regular_branches.where(:name => branch).first
+
+    # If we already have a failure record, call notify with that record
+    return branch_record.notify_of_failure if branch_record.previously_failing?
+
+    # otherwise, check if any builds exist with a failures, and if so, create a
+    # new BuildFailure record.
+    v3_client     = TravisV3Client.new(:repo => Travis::Repository.find(repo.name))
+    branch_builds = v3_client.repo_branch_builds(branch)
+
+    if branch_builds.first.failed?
+      first_failure = find_first_recent_failure(branch_builds)
+      branch_record.update(:travis_build_failure_id => first_failure.id)
+
+      branch_record.notify_of_failure
+    end
+  end
+
+  private
+
+  def find_first_recent_failure(builds)
+    first_recent_failure = nil
+    builds.each do |build|
+      if build.passed?
+        break
+      else
+        first_recent_failure = build
+      end
+    end
```

Comment:
I think you can do this a little cleaner with one of the following (assuming `builds` is Enumerable)

```ruby
builds.each_cons(2).detect { |_build, next_build| next_build.passed? }.first
# OR
builds.take_while(&:failed?).last
# OR
builds.slice_before(&:passed?).first.last
```

Reply:
The only reason I was doing it this way was to avoid multiple loop iterations (sorry, I come from the land of "performance"...), and was not able to recall some of the other iterator methods to do this in a simpler manner.

I would argue that the method name makes it clear what the intent is and since testing so far has been pretty manual for me so far, I am hesitant to make a change for something I know works currently.

That said, I will look into this, but if I push back (for now), that would be why.

Reply:
Okay, took the second option since it responded the most like the original.  Added a test as well, though it is kind of crappy since it means to be able to remove the method all together (which kinda just makes sense at this point), we would need to remove the test.

Going to leave it for now so it is merge-worthy as is, but am looking at adding more tests in a way that would allow this refactoring to happen.  Can be done as a follow up though.